src/wast-parser.cc

```diff
@@ -2771,32 +2769,33 @@ Result WastParser::ParseExpr(ExprList* exprs) {
         EXPECT(Do);
         CHECK_RESULT(ParseInstrList(&expr->block.exprs));
         EXPECT(Rpar);
-        EXPECT(Lpar);
-        TokenType type = Peek();
-        switch (type) {
-          case TokenType::Catch:
-          case TokenType::CatchAll:
-            CHECK_RESULT(ParseCatchExprList(&expr->catches));
-            expr->kind = TryKind::Catch;
-            break;
-          case TokenType::Unwind:
-            Consume();
-            CHECK_RESULT(ParseTerminatingInstrList(&expr->unwind));
-            expr->kind = TryKind::Unwind;
-            EXPECT(Rpar);
-            break;
-          case TokenType::Delegate: {
-            Consume();
-            Var var;
-            CHECK_RESULT(ParseVar(&var));
-            expr->delegate_target = var;
-            expr->kind = TryKind::Delegate;
-            EXPECT(Rpar);
-            break;
+        if (PeekMatch(TokenType::Lpar)) {
+          Consume();
+          TokenType type = Peek();
+          switch (type) {
+            case TokenType::Catch:
+            case TokenType::CatchAll:
+              CHECK_RESULT(ParseCatchExprList(&expr->catches));
+              expr->kind = TryKind::Catch;
+              break;
+            case TokenType::Unwind:
+              Consume();
+              CHECK_RESULT(ParseTerminatingInstrList(&expr->unwind));
+              expr->kind = TryKind::Unwind;
+              EXPECT(Rpar);
+              break;
+            case TokenType::Delegate: {
+              Consume();
+              Var var;
+              CHECK_RESULT(ParseVar(&var));
+              expr->delegate_target = var;
+              expr->kind = TryKind::Delegate;
+              EXPECT(Rpar);
+              break;
+            }
+            default:
+              break;
```

Comment:
We enter this `switch` only when `if (PeekMatch(TokenType::Lpar))` is true, right? Then shouldn't we still expect one of `catch` or `delegate`? If this is not an error, this sounds like we don't error out in this case? (Because of mismatched parens, it will error eventually later, but not here I guess?)
```wasm
(try
  (do ... )
  ( ;; started lpar, and didn't finish ??
```
I'm not very familiar with the code here so I might be mistaken.

Reply:
Thanks, that's a good point. I think the clause check got lost in restructuring this code. I pushed a fix and also added a new test that will exercise this case.